### PR TITLE
Multiple globe materials makes the globe darker

### PR DESCRIPTION
**Pull request by ProjectBarks**
### Fixes
- Found bug where gamma was being correct on alpha, which had no impact
- Fix that only adjusts gamma once

### Notes: 
While the gamma is adjusted per each material type I would like to make it so that it is adjusted on the final RGB values.

### Screenshots:
<h4><p align="center">Before</p></h4>

|  image |  image |
|---|---|

<h4><p align="center">After</p></h4>

|  image | image |
|---|---|

<h4><p align="center">After with HDR</p></h4>

| image | image |
|---|---|

In reference to: https://github.com/AnalyticalGraphicsInc/cesium/issues/7726

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @ProjectBarks!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/material-gamma-fix/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

**Comment by lilleyse** [comment]
How does this look with HDR enabled? It's important to still have a gamma correct somewhere if HDR is on.

**Comment by ProjectBarks** [comment]
@lilleyse updated with HDR

**Comment by mramato** [comment]
@lilleyse does anything else need to happen here?

**Comment by lilleyse** [comment]
I'll try to check this out later today.

**Comment by lilleyse** [comment]
> From what I saw, each material applies its own gamma correction so the one I removed was the duplicate in the fusion process. Let me know if that was incorrect.

This might not be true for all materials. Check out this Sandcastle example [local](http://localhost:8080/Apps/Sandcastle/index.html#c=ZZHfa8IwEMf/ldCXVqip7fBlqzLQl4EwcLKngsT0rNnyQy5p/TH835fa6ZzLQ+Duvp9v7i4NQ9II2AGSEdGwIxOwolb0/ZyLQn4OJ0Y7JjRg2HsqdKenloMGuhHVZnrQTAk+Z7oCb+OwBi+7CkE74QRYysoy+io08UeU5JGEC9i7GqEkCNx5WkJMji+6hD3JwrhTXkse+IHbw43BUmjmwPrCT9fzi5au0agpVAhgo36aZnQQk2zY3v10kNFhTB581It/DZW3QsFk2xeliVCsApt0xsuZqcxyYqRB+rGtwo46FfrUux30aIxamOhu7FYSxEFu3UHCuEOfhdoadKRGGfnHHKitbEdJVjX/BEe5PWN5coHyUjR+aaMiuPuRIiBcMmt9ZV1L+SaOUATjPPH6P5g0zK+rem0AJTu0kk06nnVJSmme+PA/5YyRK4Y3jt8) vs. [master](https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/#c=ZZHfa8IwEMf/ldCXVqip7fBlqzLQl4EwcLKngsT0rNnyQy5p/TH835fa6ZzLQ+Duvp9v7i4NQ9II2AGSEdGwIxOwolb0/ZyLQn4OJ0Y7JjRg2HsqdKenloMGuhHVZnrQTAk+Z7oCb+OwBi+7CkE74QRYysoy+io08UeU5JGEC9i7GqEkCNx5WkJMji+6hD3JwrhTXkse+IHbw43BUmjmwPrCT9fzi5au0agpVAhgo36aZnQQk2zY3v10kNFhTB581It/DZW3QsFk2xeliVCsApt0xsuZqcxyYqRB+rGtwo46FfrUux30aIxamOhu7FYSxEFu3UHCuEOfhdoadKRGGfnHHKitbEdJVjX/BEe5PWN5coHyUjR+aaMiuPuRIiBcMmt9ZV1L+SaOUATjPPH6P5g0zK+rem0AJTu0kk06nnVJSmme+PA/5YyRK4Y3jt8) where `local` doesn't apply gamma correction to the texture and it looks washed out.

**Local**
UNADJUSTEDNONRAW_thumb_a
**Master**
UNADJUSTEDNONRAW_thumb_b

**Comment by ProjectBarks** [comment]
I think I came to a good conclusion. Let me know if you find any other issues.

File: Source/Scene/Material.js [hunk 1]
```diff
@@ -703,9 +704,15 @@ define([
...
                     if (components.hasOwnProperty(component)) {
                         if (component === 'diffuse' || component === 'emission') {
-                            material.shaderSource += 'material.' + component + ' = czm_gammaCorrect(' + components[component] + '); \n';
+                            material.shaderSource += 'material.' + component + ' = ';
+                            if (isMultiMaterial) {
+                                material.shaderSource += components[component];
+                            } else {
+                                material.shaderSource += 'czm_gammaCorrect(' + components[component]  + ')';
+                            }
+                            material.shaderSource += '; \n';
```
Comment (lilleyse) [inline Source/Scene/Material.js]: Maybe slightly more concise as:

```js
                            var componentSource = isMultiMaterial ? components[component] : 'czm_gammaCorrect(' + components[component]  + ')';
                            material.shaderSource += 'material.' + component + ' = ' + componentSource + '; \n';
```

File: Source/Scene/Material.js [hunk 2]
```diff
@@ -693,6 +693,7 @@ define([
     // Create the czm_getMaterial method body using source or components.
     function createMethodDefinition(material) {
         var components = material._template.components;
+        var isMultiMaterial = material._template.materials.length > 0;
```
Comment (lilleyse) [inline Source/Scene/Material.js]: `material._template.materials` is an object so this needs to be `Object.keys(material._template.materials.length).length > 0;`

**Comment by ProjectBarks** [comment]
@lilleyse I think it should now. Now we check if the component fuses multiple sub materials

File: Source/Scene/Material.js [hunk 3]
```diff
@@ -700,12 +710,16 @@ define([
...
             if (defined(components)) {
+                var isMultiMaterial = Object.keys(material._template.materials).length > 0;
                 for ( var component in components) {
                     if (components.hasOwnProperty(component)) {
                         if (component === 'diffuse' || component === 'emission') {
-                            material.shaderSource += 'material.' + component + ' = czm_gammaCorrect(' + components[component] + '); \n';
+                            var isFusion = isMultiMaterial && isMaterialFused(components[component], material);
+                            var componentSource = isFusion ? components[component] : 'czm_gammaCorrect(' + components[component]  + ')';
+                            material.shaderSource += 'material.' + component + ' = ' + componentSource + '; \n';
+                            material.shaderSource += '; \n';
```
Comment (lilleyse) [inline Source/Scene/Material.js]: This line can be removed now.

File: Source/Scene/Material.js [hunk 4]
```diff
@@ -690,6 +690,16 @@ define([
...
 
+    function isMaterialFused(shaderComponent, material) {
+        for ( var subMaterialId in material._template.materials) {
+            if (shaderComponent.indexOf(subMaterialId) > -1) {
+                return true;
+            }
+        }
+
+        return false;
+    }
```
Comment (lilleyse) [inline Source/Scene/Material.js]: Use `hasOwnProperty` - eslint doesn't catch it but it's good practice.

```js
    function isMaterialFused(shaderComponent, material) {
        var materials = material._template.materials;
        for (var subMaterialId in materials) {
            if (materials.hasOwnProperty(subMaterialId)) {
                if (shaderComponent.indexOf(subMaterialId) > -1) {
                    return true;
                }
            }
        }

        return false;
    }
```

Comment (mramato) [reply Source/Scene/Material.js]: If it doesn't catch it, we should fix that.  https://eslint.org/docs/rules/guard-for-in is a candidate.  Can you write up an issue so we look at it?

Comment (lilleyse) [reply Source/Scene/Material.js]: Opened https://github.com/AnalyticalGraphicsInc/cesium/issues/8100.

**Comment by lilleyse** [comment]
The updates look good, thanks @ProjectBarks!

**Pull request merged**